### ISSUE TITLE
YOLO11 Segment Problems

### ISSUE THREAD

**Issue by wudiabin** [ultralytics/ultralytics#24935]
### Search before asking

- [x] I have searched the Ultralytics YOLO [issues](https://github.com/ultralytics/ultralytics/issues) and [discussions](https://github.com/orgs/ultralytics/discussions) and found no similar questions.

### Question

1. After training the YOLO11 instance segmentation model, during inference the mask might extend beyond the detection box. Is this normal, and what’s the reason for it?
2. In one of my scenarios, as shown in the picture, when annotating the image, since the right boundary has already reached the edge of the image and doesn’t include the real edges of the object, I only marked a few points. Could this kind of annotation cause issues when inferring on similar image edges later? If problems arise, what would be the reason?

<img width="1051" height="854" alt="Image" src="https://github.com/user-attachments/assets/c0a589df-4eb9-425f-82d9-36094233e346" />

3.When running inference with the YOLO11 segmentation model, the object masks and detection boxes can be incomplete. This issue happens more often at the image boundaries, similar to the right edge of the image in Problem 2. It's common for the model's predictions not to reach the far right side of the image, and the actual predicted results might be slightly to the left of the image boundary, causing parts of the detected objects to be missed.

**Comment by UltralyticsAssistant** [comment]
👋 Hello @wudiabin, thank you for your interest in Ultralytics 🚀! This is an automated response to help get your issue routed quickly, and an Ultralytics engineer will also assist you soon.
We recommend a visit to the [Docs](https://docs.ultralytics.com/) for new users where you can find many [Python](https://docs.ultralytics.com/usage/python) and [CLI](https://docs.ultralytics.com/usage/cli) usage examples and where many of the most common questions may already be answered.
If this is a 🐛 Bug Report, please provide a [minimum reproducible example](https://docs.ultralytics.com/help/minimum-reproducible-example) to help us debug it.
If this is a custom training ❓ Question, please provide as much information as possible, including dataset image examples, annotation examples, training logs, your training command, your inference command, and verify you are following our [Tips for Best Training Results](https://docs.ultralytics.com/guides/model-training-tips).
For segmentation-related questions like this 🖼️, it would also help to share:
- A few representative training images and labels
- The exact model used
- The `ultralytics` package version
- Whether the behavior appears on validation images, test images, or both
- A small reproducible sample showing the annotation pattern near image borders
Join the Ultralytics community where it suits you best. For real-time chat, head to [Discord](https://discord.com/invite/ultralytics) 🎧. Prefer in-depth discussions? Check out [Discourse](https://community.ultralytics.com/). Or dive into threads on our [Subreddit](https://www.reddit.com/r/Ultralytics/) to share knowledge with the community.
## Upgrade
Upgrade to the latest `ultralytics` package including all [requirements](https://github.com/ultralytics/ultralytics/blob/main/pyproject.toml) in a [**Python>=3.8**](https://www.python.org/) environment with [**PyTorch>=1.8**](https://pytorch.org/get-started/locally/) to verify your issue is not already resolved in the latest version:
```bash
pip install -U ultralytics
```
## Environments
YOLO may be run in any of the following up-to-date verified environments (with all dependencies including [CUDA](https://developer.nvidia.com/cuda)/[CUDNN](https://developer.nvidia.com/cudnn), [Python](https://www.python.org/) and [PyTorch](https://pytorch.org/) preinstalled):
- **Notebooks** with free GPU: <a href="https://console.paperspace.com/github/ultralytics/ultralytics"><img src="https://assets.paperspace.io/img/gradient-badge.svg" alt="Run on Gradient"/></a> <a href="https://colab.research.google.com/github/ultralytics/ultralytics/blob/main/examples/tutorial.ipynb"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"></a> <a href="https://www.kaggle.com/models/ultralytics/yolo26"><img src="https://kaggle.com/static/images/open-in-kaggle.svg" alt="Open In Kaggle"></a>
- **Google Cloud** Deep Learning VM. See [GCP Quickstart Guide](https://docs.ultralytics.com/yolov5/environments/google_cloud_quickstart_tutorial)
- **Amazon** Deep Learning AMI. See [AWS Quickstart Guide](https://docs.ultralytics.com/yolov5/environments/aws_quickstart_tutorial)
- **Docker Image**. See [Docker Quickstart Guide](https://docs.ultralytics.com/yolov5/environments/docker_image_quickstart_tutorial) <a href="https://hub.docker.com/r/ultralytics/ultralytics"><img src="https://img.shields.io/docker/pulls/ultralytics/ultralytics?logo=docker" alt="Docker Pulls"></a>
## Status
<a href="https://github.com/ultralytics/ultralytics/actions/workflows/ci.yml?query=event%3Aschedule"><img src="https://github.com/ultralytics/ultralytics/actions/workflows/ci.yml/badge.svg" alt="Ultralytics CI"></a>
If this badge is green, all [Ultralytics CI](https://github.com/ultralytics/ultralytics/actions/workflows/ci.yml?query=event%3Aschedule) tests are currently passing. CI tests verify correct operation of all YOLO [Modes](https://docs.ultralytics.com/modes) and [Tasks](https://docs.ultralytics.com/tasks) on macOS, Windows, and Ubuntu every 24 hours and on every commit.

**Comment by Y-T-G** [comment]
1. It's possible. If you use Ultralytics for inference, it will crop the masks using the boxes.
2. It can cause issue if the object is shifted right during augmentations which will result in those points removed because they move outside the image bounds. And since you only have few points there, it will break the annotation. You should add more points along the edge.
3. This can happen if your object is too big. You should use a smaller imgsz in that case because the model will have trouble regressing large objects if the imgsz is also large.

**Comment by wudiabin** [comment]
Thanks a lot for your reply!

1. I’m wondering, aren’t detection boxes supposed to crop the mask? So why do masks sometimes go beyond the detection box?
2. Does this mean that for objects near the image boundaries, it’s better to set the translate parameter to 0? Because even if I label extra points near the right edge, translation could still make some points on one side disappear.
3. About question 3, I did notice the issue you mentioned. Since my original images might be 4096×3000, I found that during segmentation model training, imgsz=1280 is more likely to cause boundary loss or object truncation than imgsz=640, but I’ve never known exactly why. Could you explain?
4. For question 3, I personally found a possible root cause: the original mask doesn’t have boundary loss, but the bounding box does, which then ends up cropping the mask, making both the detection box and mask appear truncated at the boundary. Why does this happen?

**Comment by Y-T-G** [comment]
1. It's because the model isn't penalized for masks beyond boxes during training.

2. It's only a problem if there aren't enough points. If there are enough points, then it will only remove the ones out of border, but it will still have the other points to fallback to.

3. Each grid cell has a limited receptive field which means it doesn't see the whole image. If the input size is too large, it only sees a portion of the image size. YOLO11 also uses DFL which can only regress to a fixed distance. YOLO26 doesn't have this issue.

4. During training, the loss is calculated only on the cropped mask region. So even if the model predicts masks outside, it will not be penalized. If you want to change thay behavior, you can remove the crop here:

https://github.com/ultralytics/ultralytics/blob/85c43beaab3c7fca2eaa0a17a22d742d56569327/ultralytics/utils/loss.py#L576

**Comment by yang-fei** [comment]
@Y-T-G I think the problem mentioned above is the box predicted by YOLO11 is too small sometimes. The third you said might be the true reason. YOLO11 cannot predict too large boxes, especially when the size of the object is almost half of the image. You said YOLO26 doesn't have this issue. Has it some special changes?

**Comment by Y-T-G** [comment]
YOLO26 doesn't use DFL which uses a fixed `reg_max` that limits how far the boxes can regress. But it still has receptive field limits. If you want to regress larger boxes, you would need to add an additional scale like P6:

https://github.com/Y-T-G/community/blob/main/cfg/segment/yolo11-p6.yaml

**Comment by wudiabin** [comment]
Thanks a lot for your reply. If I want to add a P6 to increase the model's receptive field, how should I train it to get a pre-trained model similar to yolo11s.pt/yolo11s-seg.pt? For example, how should I set up the dataset, training parameters, and strategies? Because after modifying P6, the original pre-trained model can no longer be used.

**Comment by Y-T-G** [comment]
You can use this to transfer the weights from pretrained model. The extra layers will have missing weights, but it shouldn't be a big concern:

https://github.com/Y-T-G/community/discussions/70#discussioncomment-14834558

**Comment by wudiabin** [comment]
<img width="1919" height="808" alt="Image" src="https://github.com/user-attachments/assets/b09d1a6a-b84f-481b-9536-8072dace9536" />

<img width="1919" height="805" alt="Image" src="https://github.com/user-attachments/assets/312a4c09-5c0a-48f7-9008-97a2163500ab" />
Figures 1 and 2 show the model training and inference results under two labeling methods (labels on the left, inference on the right)

When I was doing experiments, I found that indeed, marking more points on the boundaries can somewhat alleviate this problem. Analysis revealed that the reason is that when YOLO performs data augmentation (translation, scaling, mosaic), it applies the same geometric transformations to the polygon points, then regenerates the bbox based on the transformed points. The issue arises here: the logic for generating bboxes recalculates the bbox using the polygon points that are still inside the image after augmentation (i.e., determining the bbox boundaries based on the min/max of the annotated points). This means that when there are few boundary points, data augmentation might push some of the boundary points out, causing the polygon points to suddenly jump from the far right to the left edge. This leads to the training detection boxes shrinking, which in turn causes the inference detection boxes to shrink as well, and the cropped mask results also shrink.  

1. Is my understanding correct?  
2. If my understanding is correct, does this mean there might be a bug in YOLO's current post-processing logic? For images where the target itself reaches the image boundary, during data augmentation, if points go beyond the image, it should determine the possible boundary points using the intersection of the mask and the image boundary. In this case, the polygon points shouldn’t be used to determine the bbox boundaries, similar to how we clip bounding boxes that go beyond the target region when training a detection model. How should I modify the code for the post-processing logic that determines bbox boundaries from polygon points so that it can also handle this situation?

**Comment by glenn-jocher** [comment]
Your understanding is broadly correct: this is a training-time augmentation/label conversion edge case rather than inference post-processing. In `RandomPerspective.apply_segments()`, transformed polygons are converted back to boxes with `segment2box()`, and sparse edge annotations can lose the true image-edge extent after translate/scale/mosaic, producing smaller target boxes that later crop masks. For a local experiment, the right place to modify is [`ultralytics/utils/ops.py::segment2box()`](https://docs.ultralytics.com/reference/utils/ops#ultralytics.utils.ops.segment2box), e.g. compute min/max from clipped polygon coordinates instead of only remaining inside points; for a robust upstream fix, polygon-image boundary intersections should be inserted before bbox generation in [`RandomPerspective.apply_segments()`](https://docs.ultralytics.com/reference/data/augment#ultralytics.data.augment.RandomPerspective.apply_segments). In practice, the safest dataset-side fix is still to add enough points along image-border object boundaries so augmentation preserves the intended extent.
